tensorflow_probability/python/experimental/stats/sample_stats.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2018 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Functions for computing statistics of samples."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# Dependency imports
+
+import collections
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.internal import prefer_static as ps
+
+
+__all__ = [
+    'RunningCovariance',
+    'RunningCovarianceState',
+    'RunningVariance',
+]
+
+
+RunningCovarianceState = collections.namedtuple(
+    'RunningCovarianceState', 'num_samples, mean, comoment')
+
+
+def _update_running_covariance(
+    state, new_sample, event_ndims, dtype, axis
+):
+  new_sample = tf.cast(new_sample, dtype=dtype)
+  if axis is not None:
+    chunk_n = tf.cast(ps.shape(new_sample)[axis], dtype=dtype)
+    chunk_mean = tf.math.reduce_mean(new_sample, axis=axis)
+    chunk_delta_mean = new_sample - tf.expand_dims(chunk_mean, axis=axis)
+    chunk_comoment = tf.reduce_sum(
+        _batch_outer_product(chunk_delta_mean, event_ndims),
+        axis=axis
+    )
+  else:
+    chunk_n = tf.ones((), dtype=dtype)
+    chunk_mean = new_sample
+    chunk_comoment = tf.zeros(ps.shape(state.comoment), dtype=dtype)
+
+  new_n = state.num_samples + chunk_n
+  delta_mean = chunk_mean - state.mean
+  new_mean = state.mean + chunk_n * delta_mean / new_n
+  all_pairwise_deltas = _batch_outer_product(
+      delta_mean, event_ndims)
+  adj_factor = state.num_samples * chunk_n / (state.num_samples + chunk_n)
+  new_comoment = (state.comoment + chunk_comoment
```

Comment:
The original paper divides the comoment by n each time, meaning that it doesn't require the 'finalize' method (at least, modulo the ddof arg), and let's you easily do smoothing and moving averages by manipulating the `num_samples` state independent of the updates. Was there a reason why the `num_samples` was multiplied into the formula here?

Either way, as is `comoment` is misnamed, since it's `comoment_sum` or something like that.

Reply:
The point about simplifying without a `finalize` sounds important. Is the "original paper" the technical report I cited (https://prod-ng.sandia.gov/techlib-noauth/access-control.cgi/2008/086212.pdf) or something else? I'm not entirely clear about the ease of smoothing/moving averages. `num_samples` is in the formula because I believe that was the final proposal from the technical paper. Do you think it would be better to change the API to omit `finalize`? I wanted to support multiple choices of degrees of freedom without redoing computation (i.e. streaming computation on a bunch of samples, then ask for the sample and population covariance).

Also, I thought comoment was defined as that sum (i.e. \sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y}). Please let me know if I'm mistaken haha then the naming should definitely change.

Reply:
Looks like I was confused by the notation in that paper, there they always operate on sums rather than averages. Moments, however, are averages (or expectations) by (common?) definition, so the 'M' (and 'C') in that paper are not (co)moments. I don't know if there's a nice name for that, it's commonly described as `sum of squared deviations` in case of variance.

I still would prefer to have this implemented by maintaining a running covariance, rather (num_samples * covariance), but I don't have a reference for this being well behaved, so aside from the name of `comoment`, I retract my objections with the code.

Reply:
Sounds good!